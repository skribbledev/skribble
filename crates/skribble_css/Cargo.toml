[package]
authors = ["Ifiok Jr. <ifiokotung@gmail.com>"]
edition = "2021"
name = "skribble_css"
license = "MIT"
description = "Atomic CSS for your next frontend project"
version = "0.0.0"
homepage = "https://skribble.dev/"
documentation = "https://docs.rs/skribble_css"
repository = "https://github.com/skribbledev/skribble/tree/main/crates/skribble_css"

[lib]
crate-type = ["lib"]

[dependencies]
colorsys = "0.6.5"
heck = "0.4.0"
indexmap = { features = ["serde"], version = "1.8.0" }
indoc = "1.0.3"
readonly = "0.2.0"
regex = "1.5.4"
serde = "1.0.136"
serde_json = "1.0.79"
swc_atoms = "0.2.9"
swc_common = "0.17.8"
swc_ecmascript = { features = [
  "utils",
  "visit",
  "parser",
], version = "0.114.2" }

[dev-dependencies]
testing = "0.18.1"
walkdir = "2.3.2"
pretty_assertions = "1.1.0"
insta = "1.12.0"
