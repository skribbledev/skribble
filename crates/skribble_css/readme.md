# skribble_css

> Generate css atoms from TypeScript objects.

This library answers the question? What if tailwind was integrated into TypeScript.

## TODO

- [ ] `const { c } = await import('skribble-css')` is not supported yet.
- [ ] Scoped values generate false positives for class names.

  ```tsx
  import { c, cx } from 'skribble-css';

  {
    const c = { awesome: 'ðŸ˜¢' };
    c.awesome; // This still triggers the `ClassNameCollector` struct.
  }
  const Component = () => <div className={cx(c.md.block)} />;
  ```
