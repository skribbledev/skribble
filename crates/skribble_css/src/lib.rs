#![deny(clippy::all)]

pub use crate::error::{Error, Result};
pub use crate::generate_css::generate_css;

pub mod config;
pub mod constants;
pub mod error;
mod generate_css;
pub mod generate_typescript;
pub mod scanner;
mod utils;

#[cfg(test)]
pub(crate) mod test_utils;
