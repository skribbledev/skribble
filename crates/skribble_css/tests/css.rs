use skribble_css::{
  config::Config,
  generate_css,
  scanner::class_name_collector::{ClassNameCollector, ValidImport},
};
use std::path::PathBuf;
use swc_common::{input::StringInput, sync::Lrc, SourceMap};
use swc_ecmascript::{
  parser::{lexer::Lexer, Capturing, Parser, Syntax, TsConfig},
  visit::VisitWith,
};
use testing::fixture;

#[fixture("tests/fixtures/css/*/input.ts")]
fn test_css(input: PathBuf) {
  let config = Config::default();
  // let output = input.parent().unwrap().join("output.css");
  let source_map: Lrc<SourceMap> = Lrc::default();
  let source_file = source_map
    .load_file(&input)
    .unwrap_or_else(|_| panic!("Could not load source file {:?}", input));
  // let expected = std::fs::read_to_string(&output).expect(
  //   "Could not read output
  // // file",
  // );

  // This is where the source file should be loaded.

  // For now this will be hardcoded.
  let lexer = Lexer::new(
    Syntax::Typescript(TsConfig {
      tsx: true,
      ..Default::default()
    }),
    Default::default(),
    StringInput::from(&*source_file),
    None,
  );

  let mut parser = Parser::new_from(Capturing::new(lexer));

  let module = parser
    .parse_typescript_module()
    .expect("Failed to parse module.");

  let mut collector: ClassNameCollector = ClassNameCollector::new(
    &[
      ValidImport::new("skribble-css/client", "c"),
      ValidImport::new("@skribble-css/client", "c"),
    ],
    &config,
  );
  module.visit_with(&mut collector);

  collector.sort();

  // pretty_assertions::assert_str_eq!(
  //   generate_css(&config, &collector.get_class_names(),),
  //   expected
  // );

  insta::assert_snapshot!(
    input
      .parent()
      .unwrap()
      .file_name()
      .unwrap()
      .to_str()
      .unwrap(),
    generate_css(&config, &collector.get_class_names())
  );
}
