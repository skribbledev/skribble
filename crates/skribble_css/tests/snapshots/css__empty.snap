---
source: crates/skribble_css/tests/css.rs
assertion_line: 63
expression: "generate_css(&config, &collector.get_class_names())"

---

